zebra/zebra_vty.c

```diff
@@ -3787,54 +3787,57 @@ static int zebra_ip_config(struct vty *vty)
 	return write;
 }
 
-DEFUN (ip_zebra_import_table_distance,
-       ip_zebra_import_table_distance_cmd,
-       "ip import-table (1-252) [distance (1-255)] [route-map RMAP_NAME]",
-       IP_STR
-       "import routes from non-main kernel table\n"
-       "kernel routing table id\n"
-       "Distance for imported routes\n"
-       "Default distance value\n"
-       "route-map for filtering\n"
-       "route-map name\n")
+static int import_table_set(struct vty *vty, int afi, uint32_t table_id,
+			    const char *distance_str, const char *route_map)
 {
-	uint32_t table_id = 0;
-
-	table_id = strtoul(argv[2]->arg, NULL, 10);
 	int distance = ZEBRA_TABLE_DISTANCE_DEFAULT;
-	char *rmap =
-		strmatch(argv[argc - 2]->text, "route-map")
-			? XSTRDUP(MTYPE_ROUTE_MAP_NAME, argv[argc - 1]->arg)
-			: NULL;
-	int ret;
-
-	if (argc == 7 || (argc == 5 && !rmap))
-		distance = strtoul(argv[4]->arg, NULL, 10);
+	if (distance_str)
+		distance = strtoul(distance_str, NULL, 10);
 
 	if (!is_zebra_valid_kernel_table(table_id)) {
 		vty_out(vty,
 			"Invalid routing table ID, %d. Must be in range 1-252\n",
 			table_id);
-		if (rmap)
-			XFREE(MTYPE_ROUTE_MAP_NAME, rmap);
 		return CMD_WARNING;
 	}
 
 	if (is_zebra_main_routing_table(table_id)) {
 		vty_out(vty,
 			"Invalid routing table ID, %d. Must be non-default table\n",
 			table_id);
-		if (rmap)
-			XFREE(MTYPE_ROUTE_MAP_NAME, rmap);
 		return CMD_WARNING;
 	}
 
-	ret = zebra_import_table(AFI_IP, VRF_DEFAULT, table_id,
-				 distance, rmap, 1);
-	if (rmap)
-		XFREE(MTYPE_ROUTE_MAP_NAME, rmap);
+	return zebra_import_table(afi, VRF_DEFAULT, table_id, distance,
+				  route_map, 1);
+}
+
+DEFPY(ip_zebra_import_table_distance, ip_zebra_import_table_distance_cmd,
+      "ip import-table (1-252)$table "
+      "[distance (1-255)$distance] [route-map RMAP_NAME$rmap]",
+      IP_STR
+      "import routes from non-main kernel table\n"
+      "kernel routing table id\n"
+      "Distance for imported routes\n"
+      "Default distance value\n"
+      "route-map for filtering\n"
+      "route-map name\n")
+{
+	return import_table_set(vty, AFI_IP, table, distance_str, rmap);
+}
 
-	return ret;
+DEFPY(ipv6_zebra_import_table_distance, ipv6_zebra_import_table_distance_cmd,
+      "ipv6 import-table (1-252)$table "
+      "[distance (1-255)$distance] [route-map RMAP_NAME$rmap]",
+      IPV6_STR
+      "import routes from non-main kernel table\n"
+      "kernel routing table id\n"
+      "Distance for imported routes\n"
+      "Default distance value\n"
+      "route-map for filtering\n"
+      "route-map name\n")
+{
+	return import_table_set(vty, AFI_IP6, table, distance_str, rmap);
```

Comment:
Use `distance` and it will be an integer, no need to convert using strtoul().

Reply:
`import_table_set` needs to check if the optional distance is present, and if it is not present, a default value will be applied. `distance_str` could be NULL if distance is not specified.